Sources/PrimerSDK/Classes/Core/PCI/TokenizationService.swift

```diff
@@ -55,20 +55,72 @@ internal class TokenizationService: TokenizationServiceProtocol {
                 let settings: PrimerSettingsProtocol = DependencyContainer.resolve()
                 let state: AppStateProtocol = DependencyContainer.resolve()
                 
-                var isThreeDSEnabled: Bool = false
+                self.tokenizedPaymentMethodToken = paymentMethodToken
                 
+                var isThreeDSEnabled: Bool = false
                 if state.paymentMethodConfig?.paymentMethods?.filter({ ($0.options as? CardOptions)?.threeDSecureEnabled == true }).count ?? 0 > 0 {
                     isThreeDSEnabled = true
                 }
-                
-                self.tokenizedPaymentMethodToken = paymentMethodToken
-                
-                DispatchQueue.main.async {
-                    if case .VAULT = Primer.shared.flow.internalSessionFlow.uxMode {
-                        Primer.shared.delegate?.tokenAddedToVault?(paymentMethodToken)
+
+                /// 3DS requirements on tokenization are:
+                ///     - The payment method has to be a card
+                ///     - It has to be a vault flow
+                ///     - is3DSOnVaultingEnabled has to be enabled by the developer
+                ///     - 3DS has to be enabled int he payment methods options in the config object (returned by the config API call)
+                if paymentMethodToken.paymentInstrumentType == .paymentCard,
+                   Primer.shared.flow.internalSessionFlow.vaulted,
+                   settings.is3DSOnVaultingEnabled,
+                   paymentMethodToken.threeDSecureAuthentication?.responseCode != ThreeDS.ResponseCode.authSuccess,
+                   isThreeDSEnabled {
+                    #if canImport(Primer3DS)
+                    let threeDSService: ThreeDSServiceProtocol = ThreeDSService()
+                    DependencyContainer.register(threeDSService)
+
+                    threeDSService.perform3DS(
+                            paymentMethodToken: paymentMethodToken,
+                        protocolVersion: state.decodedClientToken?.env == "PRODUCTION" ? .v1 : .v2,
+                            sdkDismissed: { () in
```

Comment:
is it ok if this is empty?

Reply:
At the moment we don't need to do anything when the SDK is dismissed.